File: day08/day08.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Extensions.Configuration;
using Serilog;

namespace Shunty.AdventOfCode2020
{
    public class Day08 : AoCRunnerBase
    {
        public override async Task Execute(IConfiguration config, ILogger logger, bool useTestData)
        {
            var input = await GetInputLines(useTestData);
            int part1 = 0, part2 = 0, faultindex = 0;
            IList<string> patched;
            (part1, _) = RunProgram(input);
            var fault = true;
            while (fault)
            {
                (patched, faultindex) = PatchInput(input, faultindex);
                (part2, fault) = RunProgram(patched);
            }
            ShowResult(1, part1);
            ShowResult(2, part2);
        }

        private (IList<string>, int) PatchInput(IList<string> source, int startindex)
        {
            var result = new List<string>();
            var faultindex = 0;
            for (var i = 0; i < source.Count; i++)
            {
                var line = source[i];
                if (i > startindex && faultindex == 0)
                {
                    switch (line.Substring(0, 3))
                    {
                        case "nop":
                            line = line.Replace("nop", "jmp");
                            faultindex = i;
                            break;
                        case "jmp":
                            line = line.Replace("jmp", "nop");
                            faultindex = i;
                            break;
                        default:
                            break;
                    }
                }
                result.Add(line);
            }
            return (result, faultindex);
        }
        private (int acc, bool fault) RunProgram(IList<string> input)
        {
            int ip = 0, acc = 0;
            var iplog = new Dictionary<int,int>();
            while (true)
            {
                var instruction = input[ip].Split(' ')[0];
                var count = int.Parse(input[ip].Split(' ')[1]);
                if (iplog.ContainsKey(ip))
                {
                    return (acc, true);
                }
                iplog[ip] = 1;
                switch (instruction)
                {
                    case "nop":
                        ip++;
                        break;
                    case "acc":
                        acc += count;
                        ip++;
                        break;
                    case "jmp":
                        ip += count;
                        break;
                }
                if (ip >=input.Count)
                {
                    return (acc, false);
                }
            }
        }
    }
}

File: day08/day08.go
package main

import (
	"fmt"
	"log"
	"strconv"
	"strings"

	aocutils "../common"
)

func main() {
	input, err := aocutils.GetInputStrings("./day08-input.txt")
	if err != nil {
		log.Fatal(err)
	}

	p1, _ := runProgram(input)
	fmt.Println("Part 1: ", p1)

	p2, faultIndex := 0, -1
	var patchedSource []string
	for {
		patchedSource, faultIndex = patchInput(input, faultIndex)
		p2, err = runProgram(patchedSource)
		if err == nil {
			break
		}
	}
	fmt.Println("Part 2: ", p2)
}

func patchInput(input []string, lastPatchIndex int) ([]string, int) {
	result := make([]string, len(input))
	patchIndex := -1
	for i, s := range input {
		if i > lastPatchIndex && patchIndex < 0 {
			switch s[0:3] {
			case "nop":
				result[i] = strings.Replace(s, "nop", "jmp", 1)
				patchIndex = i
			case "jmp":
				result[i] = strings.Replace(s, "jmp", "nop", 1)
				patchIndex = i
			default:
				result[i] = s
			}
			continue
		}
		//...else
		result[i] = s
	}
	return result, patchIndex
}

func runProgram(input []string) (int, error) {
	ip, acc := 0, 0
	iplog := map[int]int{}

	for {
		parts := strings.Split(strings.TrimSpace(input[ip]), " ")
		instruction := parts[0]
		count, _ := strconv.Atoi(parts[1])

		if _, ok := iplog[ip]; ok {
			return acc, fmt.Errorf("Infinite loop detected at line %d", ip)
		}
		iplog[ip] = 1
		switch instruction {
		case "nop":
			ip++
		case "jmp":
			ip += count
		case "acc":
			acc += count
			ip++
		default:
			return 0, fmt.Errorf("Unhandled program instruction \"%s\" encountered at line %d", instruction, ip)
		}

		if ip >= len(input) || len(input[ip]) == 0 { // In case there's a rogue CR/LF at the end of the input file
			return acc, nil
		}
	}
}
